### DISCUSSION TITLE
Dependency in Base class doesn't appear to materialize

### DISCUSSION THREAD

**Discussion by upeshphable** [fastapi/fastapi#8685] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from abc import ABC
from fastapi import Request
from core.repos import AuditRepo
class BaseUseCase(ABC):
    def __init__(self, req: Request, audit : AuditRepo = Depends()
        self.audit = audit
        self.req = req
    def handle(self):
        self.audit.record(self.req) # This line throws AttributeError
        self.execute()

class CancelOrder(BaseUseCase):
    def __init__(self, some_other_dependency: SomeRepo = Depends()):
        super().__init__()
        self.some_other_dependency = some_other_dependency

    def execute(self):
        print("Cancelling Order")
        self.some_other_dependency.do_something() # This works
```

### Description

I'm stuck at a point where I need to use a common Dependency in the base class of all my usecases. For example, AuditTrail which needs to happen on all transactions.

To fulfil this requirement I set the audit as a dependency in the BaseUseCase class. However, when I call it on "self.audit.record()" it throws error: "AttributeError: 'Depends' object has no attribute 'record'"

All other dependencies in CancelOrder work fine.

Am I doing something wrong here?

### Operating System

macOS

### FastAPI Version

0.70.0

### Python Version

3.10.0

**Comment by upeshphable** [comment]
I saw another issue https://github.com/tiangolo/fastapi/issues/1105 which appeared to be similar but I can understand that the target function was not in the request path.

**Comment by tl24** [comment, marked as answer]
You're inheriting from a class and overriding the init method but not calling the base class init with the right arguments.  So your subclass is not implemented correctly.  I think your subclass needs to look like this:

```
class CancelOrder(BaseUseCase):
    def __init__(self, req: Request, audit : AuditRepo = Depends(), some_other_dependency: SomeRepo = Depends()):
        super().__init__(req, audit)
        self.some_other_dependency = some_other_dependency

    def execute(self):
        print("Cancelling Order")
        self.some_other_dependency.do_something() # This works
```

Of course this is rather tedious and not DRY to have to retype all of the Depends and pass along all base class arguments so I find dataclasses to be a better alternative.  Example:
```
from dataclasses import dataclass
@dataclass
class BaseUseCase(ABC):
    req: Request
    audit: AuditRepo = Depends()

    def handle(self):
        self.audit.record(self.req)
        self.execute()

@dataclass
class CancelOrder(BaseUseCase):
    some_other_dependency: SomeRepo = Depends()

    def execute(self):
        print("Cancelling Order")
        self.some_other_dependency.do_something()
```

**Comment by upeshphable** [comment]
thanks @tl24 for your response. Certainly the first approach I tried, it worked. But I rejected it for the same reason you pointed out - DRY. I will try with dataclass but that also seems like a workaround since dataclasses are supposed to be used to save state rather than business logic.

**Comment by upeshphable** [comment]
@tiangolo Do you have any insights into this please. There is another issue somebody opened and it appears either we are missing the knowledge to use dependency in base class.

Another developer opened an issue: #4681

**Comment by suspectinside** [comment]
@tiangolo +1, it should be totally awesome if you extend internal injection caps for base classes too...
may be smth like this (for Property (not constructor/init) injections):

```python
class BaseUseCase(ABC):
    @Inject # or `@Depends` m.b. with special new params for type hinting etc
    req: Request

    @Inject
    audit: AuditRepo

    def handle(self):
        self.audit.record(self.req)
        self.execute()

class CancelOrder(BaseUseCase):
    @Inject
    some_other_dependency_1: SomeRepo1

    # constructor injection - going to use `standard` Depends mechanics
    def __init__(self, some_other_dependency_2: SomeRepo2 = Depends()):
        super().__init__()
        self.some_other_dependency_2 = some_other_dependency_2

    def execute(self):
        print("Cancelling Order")
        self.some_other_dependency_1.do_something()
        self.some_other_dependency_2.do_something()

**Comment by upeshphable** [comment]
Just trying to see if there is any thought on this.

**Comment by jacksbox** [comment]
+1, watching this topic for news. 

It would be nice feature to have the Depends resolved for the parent class

**Comment by Kirill-Sakov** [comment]
This is my problem too.

```
class BaseRepository:
    def __init__(self, session: Session = Depends(get_db)):
        self.db = session
```

```
class EntityRepository(BaseRepository):
    def __init__(
            self,
            entity_name: str,
            field_name: str
    ):
        super().__init__()
        self.entity_name = entity_name
        self.field_name = field_name
```

Depends(get_db) is not resolved
